Added background checkpoint checker to make sure they're at least des…

shamsasari opened the pull request:
…erialisable

mikehearn commented on node/src/main/kotlin/net/corda/node/services/statemachine/StateMachineManager.kt:
```diff
@@ -129,14 +129,15 @@ class StateMachineManager(val serviceHub: ServiceHubInternal,
 
+    private val scheduler = FiberScheduler()
     private val mutex = ThreadBox(InnerState())
+    private val checkpointCheckerThread = if (serviceHub.configuration.devMode) Executors.newSingleThreadExecutor() else null
```
// This thread deserialises checkpoints in the background to shake out bugs in checkpoint restore.

mikehearn commented on node/src/main/kotlin/net/corda/node/services/statemachine/StateMachineManager.kt:
```diff
@@ -508,6 +509,14 @@ class StateMachineManager(val serviceHub: ServiceHubInternal,
         checkpointStorage.addCheckpoint(newCheckpoint)
         checkpointingMeter.mark()
+
+        checkpointCheckerThread?.execute {
```
Can we enable this only in "dev mode" (--dev flag passed at startup), in integration / unit tests, etc? Maybe we can enable it in MockNode? It seems wrong to have it be enabled all the time.

We could also use an explanation here about why we're doing it, what sort of problems it's meant to solve, etc.

shamsasari replied:
It is - the executor is only create in dev mode.

mikehearn replied:
Oh yes, you're right. I was expecting the check before the submission rather than using nullability but sure, that works.

Now the question is how to make sure that people use dev mode :-)

Can you add a // TODO to activate in integration test mode? I've been meaning to add one for a while.

shamsasari replied:
Dev mode is already enabled when nodes are run in integration tests. The reference.conf file has devMode enabled by default and this is picked by the driver, which is what most integration tests use. The same applies to `NodeBasedTest`.

shamsasari replied:
It's also turn on for the mock network as `testNodeConfiguration` has `devMode` set to true.

shamsasari commented:
Comments addressed.

mikehearn commented:
Actually can you mention this new feature in the docs too? One line should be sufficient.

mikehearn commented:
And whilst you're here can you make it so that checkpoints that fail to load at startup don't prevent the entire node from starting? We should have a try/catch handler that just logs the failure to resume the fiber and proceeds with the next one.

shamsasari commented:
Good catch

shamsasari commented:
The tests were failing because the checkpoint deserialisation was creating new instances of Kotlin `object`s. This was affecting the progress tracker steps which weren't being found in the progress tracker when they were set (due to being a different instance). I've got use some reflection hacks to get the Quasar Kryo to use our class resolver which fixes this issue.

shamsasari commented:
I added a check in SMM.stop to make sure no unrestorable checkpoints are created. This enables the tests to fail if such a thing occurs. This exposed an issue with one of the unit tests which was serialising a lazy instance. Long story short, the reflectasm library was trying to initialise a `java.` non-public class which failed due to access issues. I updated our `InstantiatorStrategy` to use `StdInstantiatorStrategy` in such cases.

shamsasari commented:
Back to you @mikehearn.

mikehearn commented on core/src/main/kotlin/net/corda/core/serialization/DefaultKryoCustomizer.kt:
```diff
@@ -119,4 +119,16 @@ object DefaultKryoCustomizer {
+        // is no no-arg constructor available.
+        private val defaultStrategy = Kryo.DefaultInstantiatorStrategy(fallbackStrategy)
+        override fun <T> newInstantiatorOf(type: Class<T>): ObjectInstantiator<T> {
+            // However this doesn't work for non-public classes in the java. namespace
```
Hmmm. I am unconvinced this is the right approach. Do we really need this? What is being caught in the checkpoint here? Can we just blacklist it with Kostas' blacklist code instead?

shamsasari replied:
It's not about blacklisting but rather an issue of simply being able to create certain objects.

What's being caught out here is `java.time.Ser`, which is the Java serialisation proxy for `Instant` (and the other data time classes). Kryo tries to deserialise this by first creating it using reflectasm, but reflectasm screws up because `Ser` is non-public and in `java.`. So to get around this and be able to create a `Ser` we use the more traditional InstantiatorStrategy.

Now you may be asking why are we dealing with Java serialisation proxies when we've installed Kryo Serialisers for Instant and the other date time classes. Good question. That would be because Quasar uses a custom Kryo, `ReplaceableObjectKryo`, which does this. One could argue that it shouldn't try to use `writeReplace` and `readResolve` if a Serialiser has been installed, but that I feel is separate issue to this PR.

mikehearn replied:
Mmmm. OK. We definitely need to spend more time on checkpointing soon.

shamsasari commented:
1. I've already add a TODO to DeserializeAsKotlinObjectDef to get rid of it, which I'll do after this PR
2. I'll add a TODO for that as well.

shamsasari commented:
Actually, regarding 2), it gives a nice error message of what's gone wrong and how to fix it. Blacklisting will just obscure the issue, right?

mikehearn commented:
2: yes maybe, but perhaps the blacklister code can be improved to give better error messages? After all, we may wish to give advice for all sorts of types ...
